packages/syft/src/syft/core/tensor/autograd/tensor.py

```diff
@@ -28,33 +32,50 @@
 
 @bind_protobuf
 class AutogradTensor(PassthroughTensor, PhiTensorAncestor, Serializable):
-    def __init__(self, child, requires_grad=False):
+    def __init__(self, child: AutogradTensor, requires_grad: bool = False) -> None:
         super().__init__(child)
 
         # whether to run backpropagation or not
         self.requires_grad = requires_grad
 
         # tensor gradient
-        self._grad = defaultdict(lambda: None)
+        self._grad: DefaultDict = defaultdict(lambda: None)
 
         # operation used to create this tensor (if any)
-        self._grad_fn = None
+        # TODO: Keep updating the list inside the union as they're built out
+        self._grad_fn: Optional[
+            Union[
+                autograd.backward_ops.AddOp,
+                autograd.backward_ops.SubOp,
+                autograd.backward_ops.MulOp,
+                autograd.backward_ops.DotOp,
+            ]
+        ] = None
 
         # list of ops which use this tensor
-        self.ops = list()
+        self.ops: List = list()
 
-        self.backprop_id: Optional[uuid.uuid4] = None
+        self.backprop_id: Optional[uuid.UUID] = None
 
-        self.n_backwards = Counter()
+        self.n_backwards: Counter[uuid.UUID] = Counter()
 
     @property
-    def grad(self):
+    def grad(self) -> Optional[np.ndarray]:
         if self.backprop_id not in self._grad:
             return None
         return self._grad[self.backprop_id]
 
+    # TODO: Keep updating the list inside the Union with new Ops as they're built out.
     @property
-    def grad_fn(self):
+    def grad_fn(
+        self,
+    ) -> Union[
+        None,
+        autograd.backward_ops.AddOp,
+        autograd.backward_ops.SubOp,
+        autograd.backward_ops.MulOp,
+        autograd.backward_ops.DotOp,
+    ]:
```

Comment:
Can we create a type out of these or to type by using the common parent interface? (autograd.backward_ops.Op)

Reply:
That's a great idea, I went back and implemented it in [my most recent commit](https://github.com/OpenMined/PySyft/blob/eaa3d29cb26c5caac2f1b8e35ec3b514818a097e/packages/syft/src/syft/core/tensor/autograd/tensor.py#L52) 👍 
Thanks Tudor!